src/plugins/dkommon/dkommon.cpp

```diff
@@ -103,193 +103,325 @@
-    : format(output)
-    , offsets(new size_t[__OFFSET_MAX])
+static event_response_t final_check_cb(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
 {
-    if ( !drakvuf_get_kernel_struct_members_array_rva(drakvuf, offset_names, __OFFSET_MAX, offsets) )
-        throw -1;
+    // Process only first callback call
+    auto plugin = get_trap_plugin<dkommon>(info);
+    if (plugin->is_stopping() && !plugin->done_final_analysis)
+    {
+        // Check hidden processes
+        //
+        auto temp_processes = std::move(plugin->live_processes);
+        plugin->live_processes.clear();
+        drakvuf_enumerate_processes(drakvuf, process_visitor, static_cast<void*>(plugin));
 
-    addr_t ml_rva = 0;
-    if ( !drakvuf_get_kernel_symbol_rva( drakvuf, "PsLoadedModuleList", &ml_rva) )
-        throw -1;
+        for (vmi_pid_t pid : temp_processes)
+        {
+            if (std::find(plugin->live_processes.begin(), plugin->live_processes.end(), pid) == plugin->live_processes.end())
+            {
+                print_hidden_process_information(drakvuf, info, plugin, pid);
+            }
+        }
 
-    if (!(modules_list_va = drakvuf_exportksym_to_va(drakvuf, 4, nullptr, "ntoskrnl.exe", ml_rva)))
+        // Check hidden drivers
+        //
+        auto temp_drivers = enumerate_drivers(plugin, drakvuf);
+        for (const auto& drvname : plugin->loaded_drivers)
+        {
+            if (std::find(temp_drivers.begin(), temp_drivers.end(), drvname) == temp_drivers.end())
+                print_driver_information(drakvuf, info, plugin->format, "Hidden Driver", drvname.c_str());
+        }
+
+        plugin->done_final_analysis = true;
+    }
+    return VMI_EVENT_RESPONSE_NONE;
+}
+
+dkommon::dkommon(drakvuf_t drakvuf, const void* config, output_format_t output)
+    : pluginex(drakvuf, output), format(output), offsets(new size_t[__OFFSET_MAX])
+{
+    if (!drakvuf_get_kernel_struct_members_array_rva(drakvuf, offset_names, __OFFSET_MAX, offsets))
         throw -1;
 
-    drivers_list = enumerate_drivers(this, drakvuf);
-    for (auto i: drivers_list)
-        PRINT_DEBUG("[DKOMmon] Found driver '%s'\n", i.c_str());
+    drakvuf_enumerate_processes(drakvuf, process_visitor, static_cast<void*>(this));
 
-    /* Setup trap for thread switch */
-    processes_trap.cb = check_hidden_process;
-    if (!drakvuf_add_trap(drakvuf, &processes_trap)) throw -1;
+    this->loaded_drivers = enumerate_drivers(this, drakvuf);
 
-    drivers_trap.cb = check_hidden_drivers;
-    if (!drakvuf_add_trap(drakvuf, &drivers_trap)) throw -1;
+    breakpoint_in_system_process_searcher bp;
+    if (!register_trap(nullptr, delete_process_cb, bp.for_syscall_name("PspProcessDelete")) ||
+        !register_trap(nullptr, insert_process_cb, bp.for_syscall_name("PspInsertProcess")) ||
+        !register_trap(nullptr, load_driver_cb, bp.for_syscall_name("NtLoadDriver")) ||
+        !register_trap(nullptr, unload_driver_cb, bp.for_syscall_name("NtUnloadDriver")))
```

Comment:
I kinda feel like it would be beneficial if there was some heuristic/timer based approach too. Perhaps hooking an interrupt handler or some common syscall so that the checks are run a bit more frequently?

Reply:
Alright, I guess doing the scan on these events and during shutdown is good enough. I would still like it if it was left as an option to do it on all CR3 events too though, just in case.